system-contracts/contracts/PubdataChunkPublisher.sol

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.20;
+
+import {IPubdataChunkPublisher} from "./interfaces/IPubdataChunkPublisher.sol";
+import {ISystemContract} from "./interfaces/ISystemContract.sol";
+import {L1_MESSENGER_CONTRACT, BLOB_SIZE_BYTES, EMPTY_BLOB_HASH} from "./Constants.sol";
+import {EfficientCall} from "./libraries/EfficientCall.sol";
+import {SystemContractHelper} from "./libraries/SystemContractHelper.sol";
+import {SystemLogKey} from "./Constants.sol";
+
+/**
+ * @author Matter Labs
+ * @custom:security-contact security@matterlabs.dev
+ * @notice Smart contract for chunking pubdata into the appropriate size for 4844 blobs.
+ */
+contract PubdataChunkPublisher is IPubdataChunkPublisher, ISystemContract {
+    /// @notice Chunks pubdata into pieces that can fit into blobs.
+    /// @param _pubdata The total l2 to l1 pubdata that will be sent via L1 blobs.
+    /// @dev Note: This is an early implementation, in the future we plan to support up to 16 blobs per l1 batch.
+    /// @dev We always publish 2 system logs even if our pubdata fits into a single blob. This makes processing logs on L1 easier.
+    function chunkAndPublishPubdata(bytes calldata _pubdata) external onlyCallFrom(address(L1_MESSENGER_CONTRACT)) {
+        require(_pubdata.length <= BLOB_SIZE_BYTES * 2, "pubdata should fit in 2 blobs");
+        // TODO: Update for dynamic number of blobs
+        bytes32 blob1Hash;
```

Comment:
The requirement of `_pubdata.length <= BLOB_SIZE_BYTES * 2` does not guarantee that for we have at least one blob (while the `blob2Hash == EMPTY_BLOB_HASH ? bytes32(0) : blob2Hash;` is only checked for the second one). While we could add `_pubdata.length > 0` also, I think it would be much better to have a generic algorithm here. 

For instance something like this could work (pseudocode):

```
bytes32[] blobHashes = new bytes32[](2);

// ... Copy the `totalBlobs` array as usual

for(uint i = 0; i < 2; i++) {
  uint start = BLOB_SIZE_BYTES * i;
  if (start >= _pubdata.length) {
    break;
  }
  
  bytes32 currentBlobHash;
  assembly {
    let ptr := add(totalBlobs, 0x20)
    currentBlobHash := keccak256(add(ptr, start), BLOB_SIZE_BYTES)
  }
  
  blobHashes[i] = currentBlobHash;
}

// Publish the two logs as usual.
```

IMHO it is:

- More generic, i.e. no one has to check whether the invariants that we hold for the second blob are working for the first blob also.
- You don't have to calculate unneeded blobs' hashes. While not that important in this particular case, still a good bonus.